File: axiom/discrete/sets/union_comprehension/nonoverlapping_converse_utility.php
<?php
require_once '..\..\..\..\utility.php';
$i = 0;
$txt[$i++] = "\\(\\forall_{\\substack{j \\in \\left[0; n\\right) \\setminus \\left\\{i\\right\\}}}{{x}_{i} \\cap {x}_{j} = \\emptyset}\\tag*{Eq[0]}\\)\\(\\left|{\\bigcup\\limits_{i=0}^{n - 1} {x}_{i}}\\right| = \\sum\\limits_{i=0}^{n - 1} \\left|{{x}_{i}}\\right|\\tag*{Eq[1]}\\)";
$txt[$i++] = "\\(\\left|{{x}_{0} \\cup {x}_{1}}\\right| = \\left|{{x}_{0}}\\right| + \\left|{{x}_{1}}\\right|\\tag*{Eq[2]}\\)";
$txt[$i++] = "\\(\\forall_{\\substack{j \\in \\left[0; n\\right) \\setminus \\left\\{1\\right\\}}}{{x}_{1} \\cap {x}_{j} = \\emptyset}\\tag*{Eq[3]}\\)";
$txt[$i++] = "\\({x}_{0} \\cap {x}_{1} = \\emptyset\\tag*{Eq[4]}\\)";
$txt[$i++] = "\\(\\left|{{x}_{0} \\cup {x}_{1}}\\right| = \\left|{{x}_{0}}\\right| + \\left|{{x}_{1}}\\right|\\tag*{Eq[2]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{n} {x}_{i}}\\right| = \\sum\\limits_{i=0}^{n} \\left|{{x}_{i}}\\right|\\tag*{Eq.plausible}\\)";
$txt[$i++] = "\\(\\bigcup\\limits_{i=0}^{n} {x}_{i} = {x}_{n} \\cup \\bigcup\\limits_{i=0}^{n - 1} {x}_{i}\\tag*{Eq[5]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{n} {x}_{i}}\\right| = \\left|{{x}_{n}}\\right| - \\left|{{x}_{n} \\cap \\bigcup\\limits_{i=0}^{n - 1} {x}_{i}}\\right| + \\left|{\\bigcup\\limits_{i=0}^{n - 1} {x}_{i}}\\right|\\tag*{Eq[6]}\\)";
$txt[$i++] = "\\(\\forall_{\\substack{0 \\le i < n}}{{x}_{i} \\cap {x}_{n} = \\emptyset}\\tag*{Eq[7]}\\)";
$txt[$i++] = "\\({x}_{n} \\cap \\bigcup\\limits_{i=0}^{n - 1} {x}_{i} = \\emptyset\\tag*{Eq[8]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{n} {x}_{i}}\\right| = \\left|{{x}_{n}}\\right| + \\left|{\\bigcup\\limits_{i=0}^{n - 1} {x}_{i}}\\right|\\tag*{Eq[9]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{n} {x}_{i}}\\right| = \\left|{{x}_{n}}\\right| + \\sum\\limits_{i=0}^{n - 1} \\left|{{x}_{i}}\\right|\\tag*{Eq[10]}\\)";
$txt[$i++] = "\\(\\sum\\limits_{i=0}^{n} \\left|{{x}_{i}}\\right| = \\left|{{x}_{n}}\\right| + \\sum\\limits_{i=0}^{n - 1} \\left|{{x}_{i}}\\right|\\tag*{Eq[11]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{n} {x}_{i}}\\right| = \\sum\\limits_{i=0}^{n} \\left|{{x}_{i}}\\right|\\tag*{Eq.plausible}\\)";
render(__FILE__, $txt);
?>        

File: axiom/discrete/sets/union_comprehension/nonoverlapping_converse_utility.py
from sympy.core.relational import Equality
from sympy.utility import plausible, Sum, Union, identity
from sympy.core.symbol import Symbol, dtype
from sympy import S
from sympy.concrete.expr_with_limits import Forall
from sympy.tensor.indexed import IndexedBase

from sympy.sets.sets import Interval
from sympy.core.numbers import oo
from axiom.discrete.sets import union

# given: x[i] & x[j] = {}
# |Union x[i]| = Sum |x[i]|


@plausible
def apply(given):
    assert given.is_Forall
    eq = given.function
    assert eq.is_Equality
    limits = given.limits
    assert len(limits) == 1
    (j, j_domain), *_ = limits
    assert j_domain.is_Complement
    
    n_interval, i = j_domain.args
    n = n_interval.end
    i, *_ = i.args
    
    intersection, emptyset = eq.args
    assert emptyset.is_EmptySet
    
    xi, xj = intersection.args
    if not xi.has(i):
        xi = xj
        assert xj.has(i)
        
    return Equality(abs(Union[i:0:n - 1](xi)), Sum[i:0:n - 1](abs(xi)), given=given)


from sympy.utility import check


@check
def prove(Eq):
    i = Symbol('i', integer=True)
    j = Symbol('j', integer=True)
    n = Symbol('n', domain=Interval(2, oo, integer=True))
    
    baseset = Interval(0, n, integer=True)
    domain = n.set
    assert n in baseset
    assert baseset & domain == domain

    x = IndexedBase('x', shape=(oo,), dtype=dtype.integer, finite=True)

    i_domain = Interval(0, n - 1, integer=True)
    
    given = Forall(Equality(x[i] & x[j], S.EmptySet), (j, i_domain - {i}))
    Eq << apply(given)
    
    Eq << Eq[-1].subs(n, 2).doit()
    
    Eq << Eq[0].subs(i, 1)
    
    Eq << Eq[-1].subs(j, 0)
    
    Eq << union.inclusion_exclusion_principle.apply(*Eq[-1].lhs.args).subs(Eq[-1])
    
    Eq.plausible = Eq[1].subs(n, n + 1)
    
    Eq << identity(Eq.plausible.lhs.arg).bisect(domain={n})
    
    Eq << union.inclusion_exclusion_principle.apply(*Eq[-1].rhs.args)
    
    Eq << Eq[0].subs(i, n).limits_subs(j, i)
    
    Eq << Eq[-1].union_comprehension((i, 0, n - 1))
    
    Eq << Eq[-3].subs(Eq[-1])
    
    Eq << Eq[-1].subs(Eq[1])
    
    Eq << identity(Eq.plausible.rhs).bisect(domain={n})
    
    Eq << Eq[-2].subs(Eq[-1].reversed)

    
if __name__ == '__main__':
    prove(__file__)

